fv3core/grid/generation.py

```diff
@@ -0,0 +1,2136 @@
+        return self._dy
+
+    @property
+    def dxa(self):
+        """
+        the length of the a-grid cells along the x-direction
+        """
+        if self._dx_agrid is None:
+            self._dx_agrid, self._dy_agrid = self._compute_dxdy_agrid()
+        return self._dx_agrid
+
+    @property
+    def dya(self):
+        """
+        the length of the a-grid cells along the y-direction
+        """
+        if self._dy_agrid is None:
+            self._dx_agrid, self._dy_agrid = self._compute_dxdy_agrid()
+        return self._dy_agrid
+
+    @property
+    def dxc(self):
+        """
+        the length of the c-grid cells along the x-direction
+        """
+        if self._dx_cgrid is None:
+            self._dx_cgrid, self._dy_cgrid = self._compute_dxdy_cgrid()
+        return self._dx_cgrid
+
+    @property
+    def dyc(self):
+        """
+        the length of the c-grid cells along the y-direction
+        """
+        if self._dy_cgrid is None:
+            self._dx_cgrid, self._dy_cgrid = self._compute_dxdy_cgrid()
+        return self._dy_cgrid
+
+    @property
+    def ak(self):
+        """
+        the ak coefficient used to calculate the pressure at a given k-level:
+        pk = ak + (bk * ps)
+        """
+        if self._ak is None:
+            (
+                self._ks,
+                self._ptop,
+                self._ak,
+                self._bk,
+            ) = self._set_hybrid_pressure_coefficients()
+        return self._ak
+
+    @property
+    def bk(self):
+        """
+        the bk coefficient used to calculate the pressure at a given k-level:
+        pk = ak + (bk * ps)
+        """
+        if self._bk is None:
+            (
+                self._ks,
+                self._ptop,
+                self._ak,
+                self._bk,
+            ) = self._set_hybrid_pressure_coefficients()
+        return self._bk
+
+    # TODO: can ks and ptop just be derived from ak and bk instead of being returned
+    # as part of _set_hybrid_pressure_coefficients?
+    @property
+    def ks(self):
+        """
+        the number of pure-pressure layers at the top of the model
+        also the level where model transitions from pure pressure to
+        hybrid pressure levels
+        """
+        if self._ks is None:
+            (
+                self._ks,
+                self._ptop,
+                self._ak,
+                self._bk,
+            ) = self._set_hybrid_pressure_coefficients()
+        return self._ks
+
+    @property
+    def ptop(self):
+        """
+        the pressure of the top of atmosphere level
+        """
+        if self._ptop is None:
+            (
+                self._ks,
+                self._ptop,
+                self._ak,
+                self._bk,
+            ) = self._set_hybrid_pressure_coefficients()
+        return self._ptop
+
+    @property
+    def ec1(self):
+        """
+        horizontal component of the local vector pointing to each cell center
```

Comment:
(This one is really tricky: I believe it is the 3D cartesian coordinates of the local unit vector in the x-direction at cell centers)